### PR TITLE
Attempt to remove the client singleton

### PR DESCRIPTION
**Pull request by rhymes**
Following issues #98 - #104 - #151 and #152 I decided to give it a try to start removing the singleton client.

As of now this is what I did:

- Introduced a Parse.create factory method (see @kenn's idea to create a factory method here https://github.com/adelevie/parse-ruby-client/issues/151#issue-50108705)

- Passed by dependency injection a client to all objects and class methods who called Parse.client (breaking the interface of Parse::Object.new due to ruby limitations in default arguments but I'm going to fix this, possibly by injecting the client using object.client = Parse.client(...) )

- kept Parse.client as the default if you don't pass a client object along

- had to implement a Parse.copy_client method because of the machinery to create objects at runtime that can contain other objects

My questions are:

- can we break the API? If so I propose to keep working on this and introduce some factory methods to create the various objects because the public interface - without the singleton - now it's not very handy

- can I have feedback :-) ?

### CODE REVIEW

**Comment by adelevie** [comment]
Hey @rhymes! I'll take a closer look a little bit later, but I just wanted to thank you for taking the time to make this contribution. Would you be interested in having commit access?

**Comment by kwent** [comment]
Nice ! :+1:

**Comment by rhymes** [comment]
@adelevie not really :-) Did you have a chance to look into it and form ideas on how to improve it?

**Comment by AndyObtiva** [comment]
My project team is very interested in seeing this merged too. I get this message from bundler when I try to use rhymes's version: 

parse-ruby-client at /Users/User/.rvm/gems/ruby-2.1.5@EventCastWeb/bundler/gems/parse-ruby-client-6d230accd51f did not have a valid gemspec.
This prevents bundler from installing bins or native extensions, but that may not affect its functionality.
The validation message from Rubygems was:
  ["test/test_client.rb"] are not files

It seems to work though. Thanks a lot for the fix.

**Comment by Xavdidtheshadow** [comment]
@AndyObtiva that's because `test_client.rb` was renamed to `test_client_init.rb`. If we adjust the gemspec accordingly that error will go away.

**Comment by rhymes** [comment]
@adelevie thanks for merging, though I fear that the next release is going to break some code for users! I understand you are comfortable with that.

@kwent @AndyObtiva @Xavdidtheshadow hope you guys have solid testing in place now that's been merge to master!

**Comment by adelevie** [comment]
I really wish I had more time for this project (I feel like a broken record saying this). I'm going to trust consensus here to make decisions. I'll also allow any suitable dev here to have commit and gem push access if they request it. @rhymes, can you suggest a good version number to bump to for a new release with these changes?

**Comment by rhymes** [comment]
As of now the API should not have been broken, it's just slightly cumbersome to pass along the client to each call due to the fact that my patch lacked additional factory methods apart from Parse.create

Considering this I would not call it a major new version, so something like 0.4 should suffice. If I'll find the time I'll submit a patch for the documentation in the next few days unless you can do it.

**Comment by adelevie** [comment]
I likely won't be able to add docs, so it'd be great if you could do that.

**Comment by kwent** [comment]
This is awesome guys ! Thank you for your hard work @rhymes. :+1:

**Comment by AndyObtiva** [comment]
Thanks a lot. My project is brand new, so no concern over breaking any older code APIs. Cheers. -Andy

**Comment by adelevie** [comment]
Looping in @ericcj in case he has an opinion here.

File: lib/parse/datatypes.rb [hunk 1]
```diff
@@ -58,8 +58,8 @@ def to_json(*a)
     end
 
     # Retrieve the Parse object referenced by this pointer.
-    def get
-      Parse.get @class_name, @parse_object_id if @parse_object_id
+    def get(client = nil)
+      Parse.get(@class_name, @parse_object_id, client = client) if @parse_object_id
```
Comment (ericcj) [inline lib/parse/datatypes.rb]: the client = client is unnecessary here

Comment (adelevie) [inline lib/parse/datatypes.rb]: Thanks, Eric! Do you agree, @rhymes?

**Comment by Xavdidtheshadow** [comment]
@adelevie I'd love commit access! 




As for version number, unless the changes would break existing projects we can keep it as a minor version number. 




That being said, if we wanted to go for a 1.0 in the coming months, I think it'd be doable. We'd need to finish the implementations for those last "TODO" methods and make sure everything is well documented and tested and we'll be on the road. 






~DB

**Comment by adelevie** [comment]
Thanks so much. Commit access is yours, use it wisely. A push for 1.0 is great, but I can't contribute to it. It would be your undertaking.

File: lib/parse/util.rb [hunk 2]
```diff
@@ -86,4 +88,26 @@ def Parse.object_pointer_hash(v)
       v.class_name.hash ^ v.id.hash
     end
   end
+
+
+  # NOTE: this mess is used to pass along the @client to internal objects
+  def Parse.copy_client(client, parsed_data)
```
Comment (ericcj) [inline lib/parse/util.rb]: is decorating with this better than passing the client directly through parse_json?

**Comment by ericcj** [comment]
this seems like a positive direction.  threading the client through every object is tricky business but doesn't seem dissimilar to how other object mappers do it.  e.g., rails has a connection attribute per class, etc.

File: lib/parse/datatypes.rb (hunk 1 above)
Comment (rhymes) [inline lib/parse/datatypes.rb]: @ericcj how do you pass the client down then in that case?

File: lib/parse/util.rb (hunk 2 above)
Comment (rhymes) [inline lib/parse/util.rb]: As you can see in the note I consider that method a problem and I would be more than happy with some refactoring and a different solution.

Thanks!

**Comment by rhymes** [comment]
@ericcj @Xavdidtheshadow @adelevie thanks!

My idea, if you guys want to push for a major release is to have some API like:

client = Parse.create()
new_object = client.object()
query = client.query()
etc.

so that you can avoid passing the client object everytime to the separate object. basically it would be something like:

def object(...):
  new Parse::Object(..., client = self)
end

does it make sense?

**Comment by allaire** [comment]
@rhymes @Xavdidtheshadow How are we suppose to use this new API globally in a rails app without having to create a new Parse client with the app id and api key? Would rather have something in an initializer.

**Comment by rhymes** [comment]
@allaire can you describe the use case? I'm not sure I've understood. You mean using objects without ever going to Parse.com?

**Comment by allaire** [comment]
@rhymes Sorry I wasn't clear. I mean since you dropped the Singleton, how can I use Parse in different controllers without having to do the following code every time I want to sens a push notification :

```ruby
  parseClient = Parse.create application_id: ENV["PARSE_APP_ID"],
                         api_key: ENV["PARSE_API_KEY"],
                         quiet: false
```

**Comment by rhymes** [comment]
@allaire the point of dropping the singleton is to allow the code to be used in applications that could be multithreaded. If you want to use a singleton it should be your choice as the application developer, not part of the library. Keep in mind that parse-ruby-client does not depend on Rails.

You can find out how to use a singleton in your application here http://stackoverflow.com/a/10762343/4186181

**Comment by allaire** [comment]
@rhymes I created a global variable in a initializer:

```
$parse = Parse.create application_id: ENV["PARSE_APP_ID"],
                       api_key: ENV["PARSE_API_KEY"],
                       quiet: false
```

Feels a bit dirty, but it works. I know a lot of people will use this library with Rails, so it would be cool to document the best way to include this the Rails Way.

**Comment by adelevie** [comment]
I agree with @allaire here. API changes are fine, especially to enable multi-threading, but I don't want to lose easy Rails compatibility.

**Comment by rhymes** [comment]
I reckon the best way is to use a singleton object in a Rails initializer and get it everywhere its needed from there, do you agree @adelevie ?

**Comment by adelevie** [comment]
Sure. What would that initializer code look like?

**Comment by rhymes** [comment]
@allaire @adelevie global variables are not thread safe so I wouldn't recommend using $parse either. If you really want a thread safe global variable I recommend reading this http://www.justinweiss.com/blog/2014/12/01/better-globals-with-a-tiny-activesupport-module/ 

Another solution is just to have class method in your business logic that returns the initialized parsed client and call it wherever you need it 

```
module ParseClient
  def self.get
    Parse.create application_id: ENV["PARSE_APP_ID"],
                       api_key: ENV["PARSE_API_KEY"],
                       quiet: false
  end
end

and use it like this:

```
ParseClient.get.query()

Or you can use a singleton like this http://stackoverflow.com/a/10762343/4186181

I fear though that we're beyond the scope of the library. The purpose of a library is not to implement the application logic.

For example, in the Rails app I'm using parse-ruby-client, I have multiple Parse apps so a global object won't do which is the main reason that brought me to contribute the patch in the first place.

**Comment by allaire** [comment]
@rhymes Where would you put that ParseClient module?

Also, what would the singleton initializer looks like? Since we're not using `.new`, I wonder if we can use the `initialize` method?

**Comment by rhymes** [comment]
@allaire the purpose of the library is to allow you to use an API in **your** application code.

As I showed you above there are many solutions to where to put the initialization code for the Parse client (or any other library that needs to be initialized for that matter). Since Parse is a HTTP library which is stateless you don't need a global connection object like you might need for a connection to a database. You can call Parse.create a thousand times with the same arguments and it will work all the same.

For where I would put exactly the module I would put it in my application logic, the quickest and dirtiest is to put it in a file in the lib/ directory in your Rails app, check that the lib is loaded in the config.autoload_paths in your application.rb and then just use it.

How to use the singleton is explained in the link I provided.

**Comment by allaire** [comment]
@rhymes This makes sense. I've added a parse_client.rb file in my lib/ directory like you suggested. Thanks!

**Comment by kenn** [comment]
I agree that a convenience method or at least a convention is needed, considering that not everyone needs threads or multiple instances.

I think [redis-rb](https://github.com/redis/redis-rb/blob/3.1/lib/redis.rb#L19-L25) is a great example - `Redis.current` is the default global storage of an instance. If you need thread-safety, just don't use `Redis.current` and initialize as many and store as you like.

I would suggest that we define `Parse.current` and `Parse.current=` methods, as a convention for single threaded use.

```ruby
# config/initializers/parse_client.rb
Parse.current = Parse.create application_id: ENV[...], api_key: ENV[...]

# Use it in the app code
Parse.current.query ...
```

**Comment by allaire** [comment]
great idea @kenn :+1:

**Comment by rhymes** [comment]
@kenn isn't it the same thing as the existing Parse.client class variable ? We just need to remove Parse.init and use Parse.create instead then.

**Comment by kenn** [comment]
@rhymes ah, if we have `Parse.client` you're right, it's the same thing. Then we just need README updated with the initializer example above.

**Comment by adelevie** [comment]
Solid discussion all around. I just wanted to take this moment to thank @rhymes and everyone else weighing in here. It is a delight to see my project live on and thrive without me.

**Comment by rhymes** [comment]
@kenn it would be wonderful if you yourself contributed the patch :-)

@adelevie I'm happy too, and it's wonderful to have discussions around features so we can make sure the code gets better quality having more sets of eyes on it. You did a great job by yourself!

**Comment by elioncho** [comment]
Hi guys, was this resolved?

**Comment by Xavdidtheshadow** [comment]
@elioncho, yeah, this was merged a while ago.

**Pull request merged**